File: Node/Sequelize_ORM/src/dal/repositories/genericRepository.js
class GenericRepositry {
    constructor(model) {
        this.model = model;
    }

    async getAll() {
        return await this.model.findAll();
    }

    async getById(id) {
        return await this.model.findByPk(id);
    }

    async create(data) {
        return await this.model.create(data);
    }

    async update(id, data) {
        const entity = await this.model.findByPk(id);
        return entity ? await entity.update(data) : null;
    }

    async delete(id) {
        const entity = await this.model.findByPk(id);
        return entity ? await entity.destroy() : null;
    }
}

module.exports = GenericRepositry;

File: Go/GORM_ORM/dal/repositories/genericRepository.go
package repositories

import (
	"gorm.io/gorm"
)

type GenericRepository[T any] interface {
	GetAll() ([]T, error)
	GetByID(id any) (*T, error)
	Create(entity *T) error
	Update(entity *T) error
	Delete(id any) error
}

type genericRepository[T any] struct {
	db *gorm.DB
}

func NewGenericRepository[T any](db *gorm.DB) GenericRepository[T] {
	return &genericRepository[T]{db: db}
}

func (r *genericRepository[T]) GetAll() ([]T, error) {
	var entities []T
	err := r.db.Find(&entities).Error
	return entities, err
}

func (r *genericRepository[T]) GetByID(id any) (*T, error) {
	var entity T
	err := r.db.First(&entity, id).Error
	if err != nil {
		return nil, err
	}
	return &entity, nil
}

func (r *genericRepository[T]) Create(entity *T) error {
	return r.db.Create(entity).Error
}

func (r *genericRepository[T]) Update(entity *T) error {
	return r.db.Save(entity).Error
}

func (r *genericRepository[T]) Delete(id any) error {
	var entity T
	return r.db.Delete(&entity, id).Error
}
